File: java/exercises/src/test/java/io/github/ajoz/workshop/fp/java/tests/theory/BelowSupplier.java
package io.github.ajoz.workshop.fp.java.tests.theory;

import org.junit.experimental.theories.ParameterSignature;
import org.junit.experimental.theories.ParameterSupplier;
import org.junit.experimental.theories.PotentialAssignment;

import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

public final class BelowSupplier extends ParameterSupplier {
    @Override
    public List<PotentialAssignment> getValueSources(final ParameterSignature sig) throws Throwable {
        final Below below = sig.getAnnotation(Below.class);
        return new Random().ints(
                below.limit(),
                Integer.MIN_VALUE,
                below.inclusive()
                        ? below.value() + 1
                        : below.value()
        )
                .boxed()
                .map(i -> PotentialAssignment.forValue("ints", i))
                .collect(Collectors.toList());
    }
}

File: kotlin/solutions/src/test/kotlin/io/github/ajoz/workshop/fp/kotlin/tests/theory/BelowSupplier.kt
package io.github.ajoz.workshop.fp.kotlin.tests.theory

import org.junit.experimental.theories.ParameterSignature
import org.junit.experimental.theories.ParameterSupplier
import org.junit.experimental.theories.PotentialAssignment
import java.util.Random
import java.util.stream.Collectors

class BelowSupplier : ParameterSupplier() {
    @Throws(Throwable::class)
    override fun getValueSources(sig: ParameterSignature): List<PotentialAssignment> {
        val below: Below = sig.getAnnotation(Below::class.java)
        return Random().ints(
                below.limit.toLong(),
                Integer.MIN_VALUE,
                if (below.inclusive)
                    below.value + 1
                else
                    below.value
        )
                .boxed()
                .map { i -> PotentialAssignment.forValue("ints", i) }
                .collect(Collectors.toList())
    }
}
